pandas/core/generic.py

```diff
@@ -4110,6 +4077,133 @@ def ranker(data):
+        numeric_only : boolean, default None
+            Include only float, int, boolean data. If None, will attempt to use
+            everything, then use only numeric data
+        interpolation : {'linear', 'lower', 'higher', 'midpoint', 'nearest'}
+            .. versionadded:: 0.18.0
+            This optional parameter specifies the interpolation method to use,
+            when the desired quantile lies between two data points `i` and `j`:
+
+            * linear: `i + (j - i) * fraction`, where `fraction` is the
+              fractional part of the index surrounded by `i` and `j`.
+            * lower: `i`.
+            * higher: `j`.
+            * nearest: `i` or `j` whichever is nearest.
+            * midpoint: (`i` + `j`) / 2.
+
+        Returns
+        -------
+        %s
+
+        Examples
+        --------
+
+        >>> s = Series([1, 2, 3, 4])
+        >>> s.quantile(.5)
+            2.5
+        >>> s.quantile([.25, .5, .75])
+        0.25    1.75
+        0.50    2.50
+        0.75    3.25
+        dtype: float64
+        >>> df = DataFrame(np.array([[1, 1], [2, 10], [3, 100], [4, 100]]),
+                           columns=['a', 'b'])
+        >>> df.quantile(.1)
+        a    1.3
+        b    3.7
+        dtype: float64
+        >>> df.quantile([.1, .5])
+               a     b
+        0.1  1.3   3.7
+        0.5  2.5  55.0
+        """)
+
+    @Appender(_shared_docs['quantile'] % '')
+    def quantile(self, q=0.5, axis=0, numeric_only=None,
+                 interpolation='linear'):
+        if self.ndim >= 3:
+            msg = "quantile is not implemented on on Panel or PanelND objects."
+            raise NotImplementedError(msg)
+        elif self.ndim == 1:
+            result = self.to_frame().quantile(q=q, axis=axis,
+                                              numeric_only=numeric_only,
+                                              interpolation=interpolation)
+            if not com.is_list_like(q):
+                return result.iloc[0]
+            else:
+                return result[result.columns[0]]
+
+        self._check_percentile(q)
+        per = np.asarray(q) * 100
+
+        if not com.is_list_like(per):
+            per = [per]
+            q = [q]
+            squeeze = True
+        else:
+            squeeze = False
+
+        if _np_version_under1p9:
+            if interpolation != 'linear':
+                raise ValueError("Interpolation methods other than linear "
+                                 "are not supported in numpy < 1.9")
+
+        def f(arr, per, interpolation):
+            if arr._is_datelike_mixed_type:
+                values = _values_from_object(arr).view('i8')
+            else:
+                values = arr.astype(float)
+            values = values[notnull(values)]
+            if len(values) == 0:
+                return NA
+            else:
+                if _np_version_under1p9:
+                    return _quantile(values, per)
```

Comment:
have a wrapper function (defined inside ``f``) that takes the dtype of ``arr`` and wraps the output
e.g. ``Series(result, dtype=arr.type)`` this will make the dtype correct.